### PR TITLE
DQt2: language support [windows+Mac WIP]

### PR DESCRIPTION
**Pull request by rukai**
Loads the language setting from Config/Dolphin.ini, as specified by wxLanguage enum, which I have copied over to Qt for compatibility between the two.
Due to the lack of config in dolphinQt, there is no language selection.

All languages should be supported but I have only tested, Japanese, Korean, Italian and French.

screenshot
Only the strings that happen to match those in wxWidgets dolphin are translated.

Edit:
Linux is finished.
Mac should work, but is untested.

Windows support is almost done:
I need the Languages/xx/dolphin-emu.mo files to be copied to Languages/xx/LC_MESSAGES/dolphin-emu.mo
I cant figure out how to do this with MSVC, any help would be appreciated.

I have also added Externals/gettext-0.19.4 and updated Externals/libiconv-1.14 to support windows.
Please let me know if such a change is allowed.
These new externals use a MSVC project file from here https://github.com/kahrl/gettext-msvc I dont know if its licence is compatible with dolphin or not.

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3312)

### CODE REVIEW

**Comment by spxtr** [comment]
What're the advantages of gettext vs Qt's own internationalization stuff?

**Comment by lioncash** [comment]
tbh, it's preferable to just use Qt's internationalization in the long run, since it doesn't create an unnecessary external dependency outside of the UI library itself

**Comment by rukai** [comment]
# QTranslator
I attempted to implement translations with QTranslator before gettext so I can give you a quick rundown on the options available.

We cannot solely use the Qt translation system because translations are stored in XML that requires the Qtranslator software to modify.

## Keep translations from wxWidgets
Assuming we want to keep all past and future translations from wxWidgets then an iteration would look like:
*   run xgettext - grabs strings to be translated from the source code
*   add/modify translations to gettext file
*   run lconvert - QTranslator -> gettext
*   run lrelease - compile the Qt translations

## Throwing away old translations
If we throw away old translations, an iteration would look like:
* run lupdate - grabs strings to be translated from the source code
* run lconvert - QTranslator->gettext
* add/modify translations to gettext file
* run lconvert - gettext->QTranslator
* run lrelease - compile QTranslator files

# Gettext
In the end I chose gettext because:
*   Gettext is already a requirement of wxwidgets in cmake - I know I said otherwise I was mistaken
*   Gettext is simpler then any of the Qtranslator options available
*   Gettext seems to be the standard - more people are familar with it
*   Gettext keeps translations independent of Qt
*   All the gettext infrastructure is already in place, gettext translation file locations & installation, gettextize script and transifex

# Conclusion
Gettext looks pretty good.

File: Source/Core/DolphinQt2/Translate.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <QString>
+#include <libintl.h>
+#include <locale.h>
+#include "Core/ConfigManager.h"
+#include "DolphinQt2/Translate.h"
+
+#include <iostream>
+
+void Translate::init(){
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: brace on new line

File: Source/Core/DolphinQt2/Translate.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	return QString::fromUtf8(gettext(string));
+}
+
+void Translate::setLanguage(int languageID){
+	setlocale(LC_ALL, getLocale(languageID));
+}
+
+const char* Translate::getLocale(int languageID)
+{
+	switch(languageID)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: space after `switch`

File: Source/Core/DolphinQt2/Translate.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	{
+		case wxLANGUAGE_JAPANESE: return "ja_JP.UTF-8";
+		case wxLANGUAGE_KOREAN: return "ko_KR.UTF-8";
+		case wxLANGUAGE_FRENCH: return "fr_FR.UTF-8";
+		default: return "";
+	}
+}
+
+//map enum to native language name
+const char* Translate::displayLanguage(int languageID)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: language_id

File: Source/Core/DolphinQt2/Translate.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	if(languageID != wxLANGUAGE_DEFAULT)
+		setLanguage(languageID);
+}
+
+QString Translate::translate(const char* string)
+{
+	return QString::fromUtf8(gettext(string));
+}
+
+void Translate::setLanguage(int languageID){
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: brace on newline

also language_id

File: Source/Core/DolphinQt2/Translate.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#include <iostream>
+
+void Translate::init(){
+	bindtextdomain("dolphin-emu", "/usr/local/share/locale");
+	textdomain("dolphin-emu");
+	setlocale(LC_ALL, "");
+    
+	int languageID = SConfig::GetInstance().m_InterfaceLanguage;
+	std::cout << displayLanguage(languageID) << std::endl; //testing before config UI is implemented.
+	if(languageID != wxLANGUAGE_DEFAULT)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: space after `if`

also language_id

File: Source/Core/DolphinQt2/Translate.h [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+	static void init();
+	#define _(string) Translate::translate(string)
+	static QString translate(const char* string);
+	static const char* getLocale(int languageID);
+	static const char* displayLanguage(int languageID);
+
+	enum wxLanguage
+	{
+	    /// User's default/preffered language as got from OS.
+	    wxLANGUAGE_DEFAULT,
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: using `wx` as a prefix should be dropped.

File: Source/Core/DolphinQt2/Translate.h [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QString>
+
+
+class Translate final
+{
+	public:
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: `public` should be one tab less

File: Source/Core/DolphinQt2/Translate.h [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+	static void setLanguage(int languageID);
+	static void init();
+	#define _(string) Translate::translate(string)
+	static QString translate(const char* string);
+	static const char* getLocale(int languageID);
+	static const char* displayLanguage(int languageID);
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: Function names should be CamelCase

File: Source/Core/DolphinQt2/Translate.h [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+#include <QString>
+
+
+class Translate final
+{
+	public:
+	static void setLanguage(int languageID);
+	static void init();
+	#define _(string) Translate::translate(string)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: This really shouldn't be provided as a define, and we should have the full function name specified (this is a pro of Qt's system, as it doesn't require having a define pollute every namespace).

File: Source/Core/DolphinQt2/Translate.h [hunk 10]
```diff
@@ -0,0 +1,261 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QString>
+
+
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: only one newline after headers

File: Source/Core/DolphinQt2/Translate.cpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	setlocale(LC_ALL, getLocale(languageID));
+}
+
+const char* Translate::getLocale(int languageID)
+{
+	switch(languageID)
+	{
+		case wxLANGUAGE_JAPANESE: return "ja_JP.UTF-8";
+		case wxLANGUAGE_KOREAN: return "ko_KR.UTF-8";
+		case wxLANGUAGE_FRENCH: return "fr_FR.UTF-8";
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: case statements should be aligned at the brace level

File: Source/Core/DolphinQt2/Translate.cpp [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+		default: return "";
+	}
+}
+
+//map enum to native language name
+const char* Translate::displayLanguage(int languageID)
+{
+	switch(languageID)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: case statements should be aligned at the brace level

File: Source/Core/DolphinQt2/Translate.h [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+	static void setLanguage(int languageID);
+	static void init();
+	#define _(string) Translate::translate(string)
+	static QString translate(const char* string);
+	static const char* getLocale(int languageID);
+	static const char* displayLanguage(int languageID);
+
+	enum wxLanguage
+	{
+	    /// User's default/preffered language as got from OS.
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: `preferred`

also one `/` should be removed

File: Source/Core/DolphinQt2/Translate.h [hunk 14]
```diff
@@ -0,0 +1,261 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QString>
+
+
+class Translate final
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: Since all members of this class are static, the constructor should be made private to prevent instantiation of the class.

**Comment by lioncash** [comment]
Another core issue is - how do you plan to handle translation of strings which require formatting arguments, as your current interface doesn't provide that utility, yet Qt does.

**Comment by spxtr** [comment]
I looked into Qt's internationalization a bit. Here's how it works:

Run `lupdate` on the source files. It produces a `.ts` file which is XML and looks roughly like this:
```xml
<context>
    <name>GameListModel</name>
    <message>
        <location filename="GameList/GameListModel.cpp" line="47"/>
        <source>Title</source>
        <translation type="unfinished"></translation>
    </message>
    (...)
</context>
```
Now use Qt Linguist to add a translation, item-by-item. It'll show you the code, let you save partway through, etc. It'll fill in the `<translation></translation>` parts of the `.ts` file. Run `lrelease` on it to create a binary file that allows fast lookups. Distribute this file along with the application. In `Main.cpp`, point a `QTranslator` at that file.

**Comment by rukai** [comment]
@spxtr Its my understanding that requiring Qt Linguist is bad in the same way that requiring qt Creator is bad.

File: Source/Core/DolphinQt2/Translate.h (hunk 13 above)
Comment (rukai) [inline Source/Core/DolphinQt2/Translate.h]: I copy pasted that from wx code...
Will still fix of course.

**Comment by spxtr** [comment]
I tend to agree, but there is an important difference: Qt Linguist is used by translators, who may or may not be developers. There's a good chance they use GUI translation editors, like Poedit.

**Comment by rukai** [comment]
@lioncash CharArrayFromFormatV in Source/Core/Common/StringUtil.cpp
seems to be currently used for this.

**Comment by lioncash** [comment]
The point is that it shouldn't be necessary to directly use that to get basic formatting to work, especially when a library (like Qt, or wx), provide that functionality out of the box. For example wx has wxString::Format, and Qt has tr("some string %s with %s formatters").arg("One").arg("Two"), which keep the transformed string as a wxString or QString respectively.

File: Source/Core/DolphinQt2/Translate.h [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+	static const char* GetLocale(int languageID);
+	static const char* DisplayLanguage(int languageID);
+
+	enum Language
+	{
+	    // User's default/preferred language as got from OS.
+	    LANGUAGE_DEFAULT,
+
+	    /// Unknown language, returned if wxLocale::GetSystemLanguage fails.
+	    wxLANGUAGE_UNKNOWN,
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: wx is still here

**Comment by JosJuice** [comment]
Why are you updating the po files? The internationalization bot is supposed to handle that, and most of your new strings have incorrect translations.

**Comment by spxtr** [comment]
Is it worth trying to scavenge translations from DolphinWX, or are we better off starting from scratch?

Thanks for bringing up this topic by the way, @rukai. I don't know much about this process, and it's important.

**Comment by rukai** [comment]
The internal dolphin translations and Dwx translations are stored in the same translation files.
Unless they are seperated then we cant start DQt from scratch, because we would be duping all those internal translations.

**Comment by spxtr** [comment]
Oh I see, you're right that is an issue. It looks like our options are:
* Use Qt internationalization for the UI, which will require two translation files/processes for each language, since Dolphin core uses something else.
* Use a tool like po2ts to turn the current translations into Qt ts files, scratch out anything WX, and use Qt from there.
* Use gettext for everything, which loses many advantages that Qt provides but is simpler.

That's a pickle for sure :P

**Comment by aserna3** [comment]
@dolphin-emu-bot rebuild

**Comment by rukai** [comment]
@dolphin-emu-bot rebuild

**Comment by spxtr** [comment]
I'd like to keep the Qt code... Qt. Why not use lconvert to turn .po files into .ts files at distribute time, and just use QTranslator in the code? That way all the translations stay in one place, we still get most of Qt's awesome stuff, and we don't need Translate.*.

**Comment by spxtr** [comment]
If that's not going to work, you should subclass QTranslator and implement translate there, so that we don't need to replace tr() with tra() everywhere else.

**Comment by rukai** [comment]
@spxtr I did not know I could do that, that sounds like a great option.

**Comment by rukai** [comment]
I dont agree with converting to .po files because that is messy - some parts of dolphin are translated by gettext and others translated by QTranslate.
Also its not just messy in the actual dolphin code, but in the infrastructure of ensuring the Qt translation files are distributed for all platforms, in the correct directories etc.

In a perfect world, dolphin would be written with Qt (or something else better then wx) originally, however that didnt happen and wont happen, so this looks like the best solution.

**Comment by spxtr** [comment]
Yeah, I'm okay with using gettext so long as it doesn't muck up the rest of the DQt code. I think a custom QTranslator might do the trick nicely.

**Comment by rukai** [comment]
I cant seem to get translations working as a subclass of QTranslator.
But I've also decided it would be kind of clunky. I would have to load a fake .qm file, seems a bit silly.
We can always sed /tra(/tr(/g everything if we want to move to QTranslator, if that was your concern.

**Comment by spxtr** [comment]
Why do you need to load a fake .qm file?

**Comment by rukai** [comment]
@spxtr Following the instructions here http://doc.qt.io/qt-5/qtranslator.html#details it would appear that I would need to.
However I couldnt successfully override the translate method, it couldnt seem to find the original translate method as described here http://doc.qt.io/qt-5/qtranslator.html#translate

**Comment by rukai** [comment]
At the moment I have tra() setup to take a variable number of arguments and pass these to CharArrayFromFormatV to allow for formatting arguments.
tra() returns a QString so I could very easily leave formatting arguments to Qt via args() method.

Should I keep variable args or not?

**Comment by spxtr** [comment]
I think you can just skip the load step. Lets use Qt's args stuff, and QLocale as much as possible too.

If you're having trouble finding the translate method, send me a gist or upload it on pastebin or something and I'll take a look.

**Comment by rukai** [comment]
Ok, any more problems/concerns?

File: Source/Core/DolphinQt2/GameList/GameTracker.cpp [hunk 16]
```diff
@@ -74,11 +74,11 @@ void GameTracker::GenerateFilters()
...
+		m_filters << QString::fromUtf8("*.gcm");
 	if (SConfig::GetInstance().m_ListWii || SConfig::GetInstance().m_ListGC)
-		m_filters << tr("*.iso") << tr("*.ciso") << tr("*.gcz") << tr("*.wbfs");
+		m_filters << QString::fromUtf8("*.iso") << QString::fromUtf8("*.ciso") << QString::fromUtf8("*.gcz") << QString::fromUtf8("*.wbfs");
 	if (SConfig::GetInstance().m_ListWad)
-		m_filters << tr("*.wad");
+		m_filters << QString::fromUtf8("*.wad");
 	if (SConfig::GetInstance().m_ListElfDol)
-		m_filters << tr("*.elf") << tr("*.dol");
+		m_filters << QString::fromUtf8("*.elf") << QString::fromUtf8("*.dol");
```
Comment (spxtr) [inline Source/Core/DolphinQt2/GameList/GameTracker.cpp]: These should probably be `QStringLiteral`s.

File: Source/Core/DolphinWX/CMakeLists.txt [hunk 17]
```diff
@@ -101,7 +101,6 @@ endif()
 
 set(DOLPHIN_EXE ${DOLPHIN_EXE_BASE})
 
-include(FindGettext)
```
Comment (delroth) [inline Source/Core/DolphinWX/CMakeLists.txt]: Why?

File: Source/Core/DolphinQt2/Translate.h [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QString>
+#include <QTranslator>
+
+class DQTranslator : public QTranslator
+{
+	public:
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: `public` should be at the same level of indentation as the braces also the class should be made final

File: Source/Core/DolphinQt2/Translate.h [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+	QString translate(const char * context, const char * sourceText, const char * disambiguation = 0, int n = -1) const;
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: asterisks against the type

**Comment by delroth** [comment]
To weigh in on translation mechanism, please keep to something compatible with gettext formats. We are not going to change our translation format to anything else, nor are we going to force 100+ translators to start using $random_tool_of_the_day.

File: Source/Core/DolphinQt2/Translate.h (hunk 19 above)
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: This should also have an override specifier after the `const` specifier

File: Source/Core/DolphinQt2/Translate.cpp [hunk 20]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <libintl.h>
+#include <locale.h>
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]:
```c
#include <clocale>
```

File: Source/Core/DolphinQt2/Translate.cpp [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+#include <QString>
+#include <iostream>//testing before config UI is implemented
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.cpp]: This should be removed. Tests shouldn't be left in mainline code.

File: Source/Core/DolphinQt2/Translate.h [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+	LANGUAGE_YORUBA,
+	LANGUAGE_ZHUANG,
+	LANGUAGE_ZULU,
+
+	// For custom, user-defined languages.
+	LANGUAGE_USER_DEFINED
+	};
+
+private:
+	Translate() {};
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: There's no need for the semicolon here

File: Source/Core/DolphinQt2/Translate.h [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+class DQTranslator : public QTranslator
+{
+	public:
+	QString translate(const char * context, const char * sourceText, const char * disambiguation = 0, int n = -1) const;
+};
+
+class Translate final
+{
+public:
+	static void SetLanguage(int languageID);
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: parameters should be using snake case (e.g. language_id)

File: Source/Core/DolphinQt2/Translate.h [hunk 24]
```diff
@@ -0,0 +1,270 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QString>
+#include <QTranslator>
+
+class DQTranslator : public QTranslator
```
Comment (spxtr) [inline Source/Core/DolphinQt2/Translate.h]: This can just be called Translator. I haven't been prefixing DQt stuff with DQ, so lets be consistent.

File: Source/Core/DolphinQt2/CMakeLists.txt [hunk 25]
```diff
@@ -9,6 +9,7 @@ set(SRCS
 	Host.cpp
 	RenderWidget.cpp
 	Resources.cpp
+	Translate.cpp
```
Comment (spxtr) [inline Source/Core/DolphinQt2/CMakeLists.txt]: Not a big deal, but to be consistent this should be Translator.cpp/Translator.h.

File: Source/Core/DolphinQt2/Translate.h [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+	LANGUAGE_WELSH,
+	LANGUAGE_WOLOF,
+	LANGUAGE_XHOSA,
+	LANGUAGE_YIDDISH,
+	LANGUAGE_YORUBA,
+	LANGUAGE_ZHUANG,
+	LANGUAGE_ZULU,
+
+	// For custom, user-defined languages.
+	LANGUAGE_USER_DEFINED
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translate.h]: enum contents should also be indented by one level

File: Source/Core/DolphinWX/CMakeLists.txt (hunk 17 above)
Comment (rukai) [inline Source/Core/DolphinWX/CMakeLists.txt]: It is my understanding we need gettext on all platforms, so I added it to the main CMakeLists.txt file and removed it from DolphinWX/CMakeLists.txt

Comment (spxtr) [inline Source/Core/DolphinWX/CMakeLists.txt]: Doesn't look like you checked in the main CMakeLists.txt file :P

**Comment by spxtr** [comment]
@delroth We agree with using only gettext, that's what this is.

File: Source/Core/DolphinWX/CMakeLists.txt (hunk 17 above)
Comment (rukai) [inline Source/Core/DolphinWX/CMakeLists.txt]: https://github.com/rukai/dolphin/blob/dolphinQt2/CMakeLists.txt
Looks like github doesnt show it because the file is too big but It is definitely checked in.

Comment (spxtr) [inline Source/Core/DolphinWX/CMakeLists.txt]: You're right, my bad XD

File: Source/Core/DolphinQt2/CMakeLists.txt (hunk 25 above)
Comment (rukai) [inline Source/Core/DolphinQt2/CMakeLists.txt]: Should I move all the methods from the Translate class into the Translator class?

**Comment by spxtr** [comment]
Is it possible to pull the Translate stuff into the Translator class so that we don't need both?

File: Source/Core/DolphinQt2/CMakeLists.txt (hunk 25 above)
Comment (spxtr) [inline Source/Core/DolphinQt2/CMakeLists.txt]: If possible, yes.

**Comment by rukai** [comment]
Sounds good.
since we will have a Translator object, should I keep the methods from Translate as static?

**Comment by spxtr** [comment]
I'd name the files themselves Translator.h/cpp, but it's not a big deal.

File: Source/Core/DolphinQt2/GameList/GameListModel.cpp [hunk 27]
```diff
@@ -3,6 +3,7 @@
 // Refer to the license.txt file included.
 
 #include "DolphinQt2/Resources.h"
+#include "DolphinQt2/Translator.h"
```
Comment (spxtr) [inline Source/Core/DolphinQt2/GameList/GameListModel.cpp]: Why is this needed?

**Comment by rukai** [comment]
@dolphin-emu-bot rebuild

**Comment by spxtr** [comment]
LGTM as long as moving the FindGetText to the main CMakeLists.txt is okay.

File: Source/Core/DolphinQt2/Translator.h [hunk 28]
```diff
@@ -0,0 +1,262 @@
...
+#include <QString>
+#include <QTranslator>
+
+class Translator final : public QTranslator
+{
+public:
+	QString translate(const char* context, const char* source_text, const char* disambiguation = 0, int n = -1) const override;
+	static void SetLanguage(int languageID);
+	static void Init();
+	static const char* GetLocale(int languageID);
```
Comment (spxtr) [inline Source/Core/DolphinQt2/Translator.h]: Is it possible to make this and DisplayLanguage return QString?

Comment (rukai) [inline Source/Core/DolphinQt2/Translator.h]: DisplayLanguage now returns QString.

GetLocale is only called by SetLanguage, should I merge the two?
I feel like converting the languageID into the locale string could be useful later on?

Comment (spxtr) [inline Source/Core/DolphinQt2/Translator.h]: I would make it private for now. If we need it later we can make it public again later.

File: Source/Core/DolphinQt2/Translator.h [hunk 29]
```diff
@@ -0,0 +1,262 @@
...
+#pragma once
+
+#include <QString>
+#include <QTranslator>
+
+class Translator final : public QTranslator
+{
+public:
+	QString translate(const char* context, const char* source_text, const char* disambiguation = 0, int n = -1) const override;
+	static void SetLanguage(int languageID);
```
Comment (spxtr) [inline Source/Core/DolphinQt2/Translator.h]: These should probably take a Language instead of an int, since we use another language enum elsewhere that isn't compatible.

**Comment by rukai** [comment]
@lioncash Anything else I need to work on?

File: Source/Core/DolphinQt2/Translator.cpp [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+QString Translator::translate(const char* context, const char* source_text, const char*  disambiguation, int n) const
+{
+	return QString::fromUtf8(gettext(source_text));
+}
+
+const char* Translator::GetLocale(Language language_id)
+{
+	switch (language_id)
+	{
+	case Language::CATALAN: return "ca_ES.UTF-8";
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Translator.cpp]: These should be in sorted order following the same layout as the enum, same with in DisplayLanguage.

**Comment by rukai** [comment]
Linux is finished.
Mac should work, but is untested.

Windows support is almost done:
I need the Languages/xx/dolphin-emu.mo files to be copied to Languages/xx/LC_MESSAGES/dolphin-emu.mo as well as their normal location.
I cant figure out how to do this with MSVC, any help would be appreciated.

I have also added gettext sourcecode to Externals/gettext and updated Externals/gettext
Please let me know if such a change is allowed.

**Comment by JosJuice** [comment]
The language lists in Translator.cpp need to be updated because of the changes in PR #3857.

**Comment by rukai** [comment]
I'll deal with this when dolphinQt is in a more complete state. (This work will provide a good starting point for then)
